### ISSUE TITLE
第四章 Sequential的三种写法，网络输出结果不一样

### ISSUE THREAD

**Issue by djshu** [chenyuntc/pytorch-book#71]
**第四章notebook中代码如下：**
```python
# Sequential的三种写法
net1 = nn.Sequential()
net1.add_module('conv', nn.Conv2d(3, 3, 3))
net1.add_module('batchnorm', nn.BatchNorm2d(3))
net1.add_module('activation_layer', nn.ReLU())

net2 = nn.Sequential(
        nn.Conv2d(3, 3, 3),
        nn.BatchNorm2d(3),
        nn.ReLU()
        )

from collections import OrderedDict
net3= nn.Sequential(OrderedDict([
          ('conv1', nn.Conv2d(3, 3, 3)),
          ('bn1', nn.BatchNorm2d(3)),
          ('relu1', nn.ReLU())
        ]))
print('net1:', net1)
print('net2:', net2)
print('net3:', net3)


input = V(t.rand(1, 3, 4, 4))
output = net1(input)
output = net2(input)
output = net3(input)
output = net3.relu1(net1.batchnorm(net1.conv(input)))
``` 

**我稍微修改了一下打印出网络的输出值**
```python
input = V(t.rand(1, 3, 4, 4))
input2=input
input3=input
input4=input
output1 = net1(input)
output2 = net2(input2)
output3 = net3(input3)
output4 = net3.relu1(net1.batchnorm(net1.conv(input4)))
print('input: ',input)
print('output1:',output1)
print('input2: ',input2)
print('output2:',output2)
print('input3: ',input3)
print('output3:',output3)
print('input4: ',input4)
print('output4:',output4)
```
**输出：**
```
input:  tensor([[[[ 0.8346,  0.9514,  0.6958,  0.7501],
          [ 0.7408,  0.5216,  0.2121,  0.7931],
          [ 0.7457,  0.1186,  0.5825,  0.5945],
          [ 0.0480,  0.9939,  0.9572,  0.9339]],

         [[ 0.3144,  0.7812,  0.5537,  0.7115],
          [ 0.1510,  0.1935,  0.4596,  0.4351],
          [ 0.8608,  0.8190,  0.5364,  0.0421],
          [ 0.5160,  0.5999,  0.3452,  0.7298]],

         [[ 0.6361,  0.7464,  0.5599,  0.7679],
          [ 0.8955,  0.1929,  0.3311,  0.6545],
          [ 0.0679,  0.0560,  0.1533,  0.9858],
          [ 0.3074,  0.6287,  0.2402,  0.1395]]]])
output1: tensor([[[[ 0.2010,  0.0000],
          [ 0.0000,  1.0389]],

         [[ 0.0000,  0.1681],
          [ 0.0000,  0.4952]],

         [[ 0.0000,  0.0936],
          [ 0.9633,  0.3430]]]])
input2:  tensor([[[[ 0.8346,  0.9514,  0.6958,  0.7501],
          [ 0.7408,  0.5216,  0.2121,  0.7931],
          [ 0.7457,  0.1186,  0.5825,  0.5945],
          [ 0.0480,  0.9939,  0.9572,  0.9339]],

         [[ 0.3144,  0.7812,  0.5537,  0.7115],
          [ 0.1510,  0.1935,  0.4596,  0.4351],
          [ 0.8608,  0.8190,  0.5364,  0.0421],
          [ 0.5160,  0.5999,  0.3452,  0.7298]],

         [[ 0.6361,  0.7464,  0.5599,  0.7679],
          [ 0.8955,  0.1929,  0.3311,  0.6545],
          [ 0.0679,  0.0560,  0.1533,  0.9858],
          [ 0.3074,  0.6287,  0.2402,  0.1395]]]])
output2: tensor([[[[ 0.8605,  0.5681],
          [ 0.0000,  0.0000]],

         [[ 0.0000,  0.0000],
          [ 0.8663,  0.1599]],

         [[ 0.0000,  0.5719],
          [ 0.0000,  0.9679]]]])
input3:  tensor([[[[ 0.8346,  0.9514,  0.6958,  0.7501],
          [ 0.7408,  0.5216,  0.2121,  0.7931],
          [ 0.7457,  0.1186,  0.5825,  0.5945],
          [ 0.0480,  0.9939,  0.9572,  0.9339]],

         [[ 0.3144,  0.7812,  0.5537,  0.7115],
          [ 0.1510,  0.1935,  0.4596,  0.4351],
          [ 0.8608,  0.8190,  0.5364,  0.0421],
          [ 0.5160,  0.5999,  0.3452,  0.7298]],

         [[ 0.6361,  0.7464,  0.5599,  0.7679],
          [ 0.8955,  0.1929,  0.3311,  0.6545],
          [ 0.0679,  0.0560,  0.1533,  0.9858],
          [ 0.3074,  0.6287,  0.2402,  0.1395]]]])
output3: tensor([[[[ 0.0000,  0.0000],
          [ 0.7801,  0.1952]],

         [[ 0.0464,  0.2840],
          [ 0.0927,  0.0000]],

         [[ 0.3540,  0.0000],
          [ 0.1295,  0.0000]]]])
input4:  tensor([[[[ 0.8346,  0.9514,  0.6958,  0.7501],
          [ 0.7408,  0.5216,  0.2121,  0.7931],
          [ 0.7457,  0.1186,  0.5825,  0.5945],
          [ 0.0480,  0.9939,  0.9572,  0.9339]],

         [[ 0.3144,  0.7812,  0.5537,  0.7115],
          [ 0.1510,  0.1935,  0.4596,  0.4351],
          [ 0.8608,  0.8190,  0.5364,  0.0421],
          [ 0.5160,  0.5999,  0.3452,  0.7298]],

         [[ 0.6361,  0.7464,  0.5599,  0.7679],
          [ 0.8955,  0.1929,  0.3311,  0.6545],
          [ 0.0679,  0.0560,  0.1533,  0.9858],
          [ 0.3074,  0.6287,  0.2402,  0.1395]]]])
output4: tensor([[[[ 0.2010,  0.0000],
          [ 0.0000,  1.0389]],

         [[ 0.0000,  0.1681],
          [ 0.0000,  0.4952]],

         [[ 0.0000,  0.0936],
          [ 0.9633,  0.3430]]]])
```

**output = net1(input) 和 output = net3.relu1(net1.batchnorm(net1.conv(input)))结果一样，net2和net3的输出不一样，虽然由于输入是rand出来的随机值，每次运行每个网络的输出不一样，但是4个输出之间的区别不变，请问这是什么问题？**

**Comment by djshu** [comment]
发现原因了，output1和output4输出总一样的原因是，输出output4就是使用的net1的卷积层和batchnorm层，参数一样，且relu层无参数。而net2和net3 与 net1 卷积层和batchnorm层的参数是随机生成的，不一样，所以输出不一样
